File: src/util/doctor.ts
import fs from 'fs'
import path from 'path'
import os from 'os'
import { TomlParser } from './toml.js'
import { RegistryClient } from '../registry/client.js'
import { readRc } from './keys.js'

export type CheckStatus = 'pass' | 'fail' | 'warn'

export interface CheckResult {
  name: string
  status: CheckStatus
  message: string
  fix?: string
}

export class Doctor {
  private results: CheckResult[] = []

  async runAll(): Promise<CheckResult[]> {
    await this.checkRegistry()
    this.checkAuth()
    this.checkProject()
    await this.checkDependencies()
    await this.checkNvidiaApi()
    return this.results
  }

  private addResult(name: string, status: CheckStatus, message: string, fix?: string) {
    this.results.push({ name, status, message, fix })
  }

  private async checkRegistry() {
    try {
      const client = new RegistryClient()
      await client.fetchIndex()
      this.addResult('Registry', 'pass', 'reachable')
    } catch (e: any) {
      this.addResult('Registry', 'fail', `unreachable: ${e.message}`, 'Check your network connection and QUILL_REGISTRY / LECTERN_REGISTRY env var.')
    }
  }

  private checkAuth() {
    const envToken = process.env['QUILL_TOKEN']
    const rc = readRc()
    const hasToken = !!(envToken || rc?.token)

    if (hasToken) {
      this.addResult('Auth', 'pass', 'token found')
    } else {
      this.addResult('Auth', 'warn', 'no token found (run `quill login` to publish)', 'Run `quill login` to authenticate with the registry.')
    }
  }

  private checkProject() {
    const projectDir = process.cwd()
    const tomlPath = path.join(projectDir, 'ink-package.toml')

    if (!fs.existsSync(tomlPath)) {
      this.addResult('ink-package.toml', 'warn', 'not in a project directory', 'Run `quill init` or `quill new <name>` to create a project.')
      return
    }

    try {
      TomlParser.read(tomlPath)
      this.addResult('ink-package.toml', 'pass', 'valid')
    } catch {
      this.addResult('ink-package.toml', 'fail', 'parse error', 'Check ink-package.toml for syntax errors.')
    }
  }

  private async checkDependencies() {
    const projectDir = process.cwd()
    const lockPath = path.join(projectDir, 'quill.lock')

    if (!fs.existsSync(lockPath)) {
      this.addResult('Dependencies', 'pass', 'no dependencies')
      return
    }

    try {
      const lock = JSON.parse(fs.readFileSync(lockPath, 'utf8'))
      const deps = lock.packages ? Object.keys(lock.packages) : Object.keys(lock)

      if (deps.length === 0) {
        this.addResult('Dependencies', 'pass', 'no dependencies')
        return
      }

      const client = new RegistryClient()
      const index = await client.fetchIndex()

      let allFound = true
      for (const dep of deps) {
        // Handle both Map and Proxy returned by fetchIndex
        let found = false
        if (index instanceof Map) {
          found = index.has(dep)
        } else {
          found = dep in (index as any)
        }
        if (!found) {
          allFound = false
          break
        }
      }

      if (allFound) {
        this.addResult('Dependencies', 'pass', 'all installed')
      } else {
        this.addResult('Dependencies', 'warn', 'some deps not found in registry', 'Packages may have been unlisted. Run `quill install` to reinstall.')
      }
    } catch {
      this.addResult('Dependencies', 'warn', 'could not check')
    }
  }

  private async checkNvidiaApi() {
    try {
      const res = await fetch('https://api.nvcf.nvidia.com/v2/info', {
        method: 'HEAD',
        signal: AbortSignal.timeout(5000)
      })
      if (res.ok) {
        this.addResult('NVIDIA API', 'pass', 'reachable')
      } else {
        this.addResult('NVIDIA API', 'warn', 'unreachable', 'Set NVIDIA_API_KEY env var for semantic search features.')
      }
    } catch {
      this.addResult('NVIDIA API', 'warn', 'unreachable (search features may not work)', 'Set NVIDIA_API_KEY env var for semantic search features.')
    }
  }

  printResults() {
    const maxNameLen = Math.max(...this.results.map(r => r.name.length), 10)
    const indent = '  '

    console.log('Doctor check results:')
    for (const r of this.results) {
      const icon = r.status === 'pass' ? '✓' : r.status === 'fail' ? '✗' : '⚠'
      const name = r.name.padEnd(maxNameLen)
      console.log(`${indent}${icon} ${name}  ${r.message}`)
      if (r.fix && r.status !== 'pass') {
        console.log(`${indent}  → ${r.fix}`)
      }
    }

    const passed = this.results.filter(r => r.status === 'pass').length
    const warnings = this.results.filter(r => r.status === 'warn').length
    const failed = this.results.filter(r => r.status === 'fail').length
    console.log(`\n${this.results.length} checks, ${passed} passed, ${warnings} warning, ${failed} failed`)
  }

  hasFailed(): boolean {
    return this.results.some(r => r.status === 'fail')
  }
}

File: quill/src/commands/doctor.rs
use async_trait::async_trait;

use crate::commands::Command;
use crate::context::Context;
use crate::error::Result;
use crate::registry::auth::QuillRc;
use crate::registry::RegistryClient;
use crate::util::compiler::resolve_compiler;

pub struct Doctor;

#[async_trait]
impl Command for Doctor {
    async fn execute(&self, ctx: &Context) -> Result<()> {
        println!("Running diagnostics...\n");

        let mut all_ok = true;

        // Check compiler available
        println!("Checking compiler...");
        match resolve_compiler() {
            Ok(path) => println!("  [OK] Compiler found: {}", path.display()),
            Err(_) => {
                println!("  [WARN] Compiler not found");
                println!("         Install with: quill doctor --install-compiler");
                all_ok = false;
            }
        }

        // Check registry reachable
        println!("\nChecking registry...");
        let client = RegistryClient::new(&ctx.registry_url);
        match client.fetch_index().await {
            Ok(_) => println!("  [OK] Registry reachable: {}", ctx.registry_url),
            Err(e) => {
                println!("  [WARN] Registry unreachable: {}", e);
                all_ok = false;
            }
        }

        // Check logged in state
        println!("\nChecking authentication...");
        match QuillRc::load() {
            Ok(rc) => {
                println!("  [OK] Logged in as: {}", rc.username);
                println!("       Registry: {}", rc.registry);
            }
            Err(crate::error::QuillError::NotLoggedIn) => {
                println!("  [INFO] Not logged in (this is fine for public packages)");
            }
            Err(e) => {
                println!("  [WARN] Auth error: {}", e);
            }
        }

        // Check manifest valid
        println!("\nChecking manifest...");
        if let Some(ref manifest) = ctx.manifest {
            println!("  [OK] Manifest valid");
            println!("       Package: {}@{}", manifest.package.name, manifest.package.version);
        } else {
            let manifest_path = ctx.project_dir.join("ink-manifest.toml");
            if manifest_path.exists() {
                println!("  [WARN] Manifest exists but failed to parse");
                all_ok = false;
            } else {
                println!("  [INFO] No manifest in current directory");
            }
        }

        // Check lockfile
        println!("\nChecking lockfile...");
        if let Some(ref lockfile) = ctx.lockfile {
            println!("  [OK] Lockfile valid");
            println!("       {} packages locked", lockfile.packages.len());
        } else {
            let lockfile_path = ctx.project_dir.join("quill.lock");
            if lockfile_path.exists() {
                println!("  [WARN] Lockfile exists but failed to parse");
            } else {
                println!("  [INFO] No lockfile (run 'quill install')");
            }
        }

        println!();
        if all_ok {
            println!("All checks passed!");
        } else {
            println!("Some checks failed. See above for details.");
        }

        Ok(())
    }
}
